### DISCUSSION TITLE
default ipxe script from local filsystem or over tftp/http etc..

### DISCUSSION THREAD

**Discussion by tpham3783** [ipxe/ipxe#579] [Q&A]
Is there a way to tell ipxe where to find the ipxe boot script without having to recompile the firmware?

Ideally, it would be nice to be able to specify it as a parameter from a dhcpd daemon.

thanks.

**Comment by NiKiZe** [comment]
Just send the filename to use with the DHCP response, see https://ipxe.org/howto/chainloading#breaking_the_loop_with_the_dhcp_server

You can use any DHCP option you want with an embedded script.

How do you du it today? Or what is the reason for your question?

**Reply by tpham3783** [reply]
I am not booting directly from a ipxe supported NIC.

I am booting from pxe with UEFI support to load ipxe.efi.  Since ipxe.efi was custom built with an embed script to chain load another ipxe script from the same PXE/tftp-server, I am wondering why can't we make that a default policy to always pull ipxe.conf from the same tftpserver that ipxe.efi was loaded from.  I know that ipxe is very robust and shouldn't be customized for each use case.  Is it possible to do it for the above scenario?

**Reply by NiKiZe** [reply]
Then just have an embedded script that does DHCP and chain ipxe.ipxe. note iPXE uses ipxe scripts, not configs.

Recently there is also autoexec.ipxe

But really, configure your DHCP server to send the correct filename for the correct stage, see above mentioned documentation. Also example at https://ipxe.org/appnote/proxydhcp

**Comment by tpham3783** [comment]
Nikize,

I think this feature was documented in the autoexec.ipxe script as mentioned in the link below:

 https://github.com/ipxe/ipxe/discussions/441

why can't we make it a default script, w/ a name like startup.ipxe.  This should be implicable for USB and network boot modes.

**Reply by NiKiZe** [reply]
Why do you need to have it be named startup.ipxe instead of autoexec.ipxe?
It should work for efi just fine. But I think you have misunderstood some things, and reading the linked documentation a few times might be helpful, the answer you seem to actually be seeking is in there.

Many have one ipxe binary that is never updated, while menues etc are updated all the time, even script locations. Others are using the prebuilt binaries from boot.ipxe.org no embedded script and using their own filenames.

**Reply by tpham3783** [reply]
> Why do you need to have it be named startup.ipxe instead of autoexec.ipxe? 
    uefi uses startup.nsh file as a fallback script if the boot binary could not be found at efi/boot/bootx64.efi
    I am fine with autoexec,ipxe, it is just that I couldn't fine any documentation on it until I browsed this discussion board.

> It should work for efi just fine. But I think you have misunderstood some things, and reading the linked documentation a few times might be helpful, the answer you seem to actually be seeking is in there.

  Yeah, I saw it, the magic was leasing the second response from DHCPD server to send the second stage boot-script.



> Many have one ipxe binary that is never updated, while menues etc are updated all the time, even script locations. Others are using the prebuilt binaries from boot.ipxe.org no embedded script and using their own filenames.

I see, my problem was that I didn't know that the boot-script can be loaded from:

1.  the default file: autoexec.ipxe
2.  or configured from dhcp server using the user-class option of dhcp.  ie:
```
    if exists user-class and option user-class = "iPXE" {
      filename "http://my.web.server/real_boot_script.php";
  }
```
